### PR TITLE
3 New Columns - Session Type, Recording & Chat Links, Menu Help Updated & Guest List Visibilty changed to true

### PR DESCRIPTION
**Pull request by RamMohanMadhamanchi**
- On Request from Mentor Support Team -> Guest List Visibilty changed to true (Event attendees can see guest list)
- Added 2 New Columns 
  - Recording Links
  - Chat Links
- Corresponding Help Section Updated (about new columns)
- New Button "Get recording and chat links" is added in Menu (a new function to fetch them is added)

### CODE REVIEW

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 1]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+    showInfo(TITLE_ERROR, MSG_CALENDAR_REJECTED);
+    return;
+  }
+
+  let activeSheet = SpreadsheetApp.getActiveSheet();
+  let oneToOneData = get1to1Data(activeSheet);
+  if (!oneToOneData) {
+    showInfo(TITLE_ERROR, MSG_COULD_NOT_FETCH_DATA);
+    return;
+  }
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: This block of code appears in multiple places. Please refactor this into a separate function.
```js
function initSetup() {
  let result = {
    success: false,
    calendar: null,
    activeSheet: null,
    oneToOneData: null
  };
  result.calendar = getCalendar();
  if (!result.calendar) {
    showInfo(TITLE_ERROR, MSG_CALENDAR_NOT_FOUND);
    return result;
  }

  if (!confirmCalendar(result.calendar)) {
    showInfo(TITLE_ERROR, MSG_CALENDAR_REJECTED);
    return result;
  }

  result.activeSheet = SpreadsheetApp.getActiveSheet();
  result.oneToOneData = get1to1Data(result.activeSheet);
  if (!result.oneToOneData) {
    showInfo(TITLE_ERROR, MSG_COULD_NOT_FETCH_DATA);
    return result;
  }
  result.success = true;
  return result;
}
```

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 2]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+  let noRecordingLinks = [];
+  let numSuccess = 0;
+  let numSkipped = 0;
+  let calendarId = calendar.getId();
+
+  for (let rowIndex = 0; rowIndex < oneToOneData.length; ++rowIndex) {
+    let rowNumber = rowIndex + DATA_BEGIN_ROW;
+
+    let recordingLinkCell = activeSheet.getRange(rowNumber, RECORDING_LINK_INDEX + 1);
+    let chatLinkCell = activeSheet.getRange(rowNumber, CHAT_LINK_INDEX + 1);
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Move these declarations near to their usage.

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 3]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+    let eventAttachments = eventResource.attachments;
+    
+    if (!eventAttachments) {
+      recordingLinkCell.setValue(NOT_FOUND);
+      chatLinkCell.setValue(NOT_FOUND);
+      noRecordingLinks.push(rowNumber);
+      continue;
+    } else {
+      for (let i = 0; i < eventAttachments.length; ++i) {
+        let linkUrl = eventAttachments[i].fileUrl;
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Please add null check for `linkUrl`.

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 4]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+        let queryIndex = linkUrl.indexOf("?");
+        linkUrl = linkUrl.substring(0,queryIndex);
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Can you give an example of `linkUrl`? Why are you extracting the part after `?`?

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 5]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+        if (eventAttachments[i].mimeType === "video/mp4") {
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Please use constants instead of hardcoded strings.

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 6]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+          recordingLinkCell.setValue(linkUrl);
+        }
+        else if (eventAttachments[i].mimeType === "text/plain") {
+          chatLinkCell.setValue(linkUrl);
+        }
+      }
+      ++numSuccess
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: If the video and chat are not found, you are still incrementing `numSuccess`. Use 2 booleans `videoFound`, `chatFound` and increment `numSuccess` only if both are true. This will also help you avoid the two extra `if`s below.

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 7]
```diff
@@ -626,6 +633,133 @@ function sendInvitesToMentors() {
...
+      continue;
+    } else {
+      for (let i = 0; i < eventAttachments.length; ++i) {
+        let linkUrl = eventAttachments[i].fileUrl;
+        let queryIndex = linkUrl.indexOf("?");
+        linkUrl = linkUrl.substring(0,queryIndex);
+        if (eventAttachments[i].mimeType === "video/mp4") {
+          recordingLinkCell.setValue(linkUrl);
+        }
+        else if (eventAttachments[i].mimeType === "text/plain") {
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Move `else if` into previous line - like `} else if (...`

**Review by sailesh-ramayanam — CHANGES_REQUESTED** [review]

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 4 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Example Link: https://drive.google.com/file/d/19V1RKGDaKg7FbZpddpg9jM71IcsjqZL9/view?usp=drive_web
after `?`
we can have any one of the following:
`usp=docs_home`
`usp=drivesdk`
`usp=sharing`
`usp=drive_web`
`usp=direct_url`
or may be more.
It seems to be providing data from where we got access to the document.

If you don't want me to remove it. Please let me know - I shall keep it.

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 1 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: `initSetup()` added and required changes were made

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 2 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Declarations moved near to their respective usage

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 3 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Check for `linkUrl` and `queryIndex` were added

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 5 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: `const RECORDING_MIME_TYPE = "video/mp4";`
`const CHAT_MIME_TYPE = "text/plain";`
`const QUERY_DELIMITER = "?";`

were added and used instead of hardcoded strings

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 7 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: updated format

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 6 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Requirements were added

**Comment by RamMohanMadhamanchi** [comment]
- All the requested changes were updated, please review them

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 8]
```diff
@@ -355,44 +393,31 @@ function addAttendees(calendarId, meetId, emails) {
...
+  let setupResult = initSetup();
+  if (!setupResult.success) {
     return;
   }
 
   let errorRows = [];
   let skippedRows = [];
   let successRows = [];
-  for (let rowIndex = 0; rowIndex < oneToOneData.length; ++rowIndex) {
+  for (let rowIndex = 0; rowIndex < setupResult.oneToOneData.length; ++rowIndex) {
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Please use destructuring like below. That way you don't need to edit the usages.
```js
let { calendar, activeSheet, oneToOneData } = setupResult;
```

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 9]
```diff
@@ -626,43 +625,161 @@ function sendInvitesToMentors() {
...
+      chatLinkCell.setValue(NOT_FOUND);
+      noRecordingLinks.push(rowNumber);
+      noChatLinks.push(rowNumber);
+      continue;
+    } else {
+      let videoFound = false;
+      let chatFound = false;
+      for (let i = 0; i < eventAttachments.length; ++i) {
+        let currentAttachment = eventAttachments[i];
+        let linkUrl = currentAttachment.fileUrl;
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Please use `return early` pattern - https://medium.com/swlh/return-early-pattern-3d18a41bba8 .
```js
if (!currentAttachment || !currentAttachment.fileUrl) {
    continue;
}
```

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 10]
```diff
@@ -626,43 +625,161 @@ function sendInvitesToMentors() {
...
+        ++numSuccess;
+        continue;
+      }
+      if (!videoFound && chatFound) {
+        recordingLinkCell.setValue(NOT_FOUND);
+        noRecordingLinks.push(rowNumber);
+      } else if (videoFound && !chatFound) {
+        chatLinkCell.setValue(NOT_FOUND);
+        noChatLinks.push(rowNumber);
+      }
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: It is sufficient to check for `!videoFound`. Also, no need of `else if`.
```js
      if (!videoFound) {
        recordingLinkCell.setValue(NOT_FOUND);
        noRecordingLinks.push(rowNumber);
      }
      if (!chatFound) {
        chatLinkCell.setValue(NOT_FOUND);
        noChatLinks.push(rowNumber);
      }
```

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 11]
```diff
@@ -626,43 +625,161 @@ function sendInvitesToMentors() {
...
+    let recordingLinkCell = setupResult.activeSheet.getRange(rowNumber, RECORDING_LINK_INDEX + 1);
+    let chatLinkCell = setupResult.activeSheet.getRange(rowNumber, CHAT_LINK_INDEX + 1);
+
+    if (!eventAttachments) {
+      recordingLinkCell.setValue(NOT_FOUND);
+      chatLinkCell.setValue(NOT_FOUND);
+      noRecordingLinks.push(rowNumber);
+      noChatLinks.push(rowNumber);
+      continue;
+    } else {
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: No need of `else` here. The `continue` statement in the `if` block will ensure that this code won't be executed.

**Review by sailesh-ramayanam — CHANGES_REQUESTED** [review]

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 4 above)
Comment (sailesh-ramayanam) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: @RamMohanMadhamanchi I don't see any harm in having the url as is. Please store it as it is.

Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Removed `queryIndex` and it's corresponding changes to `linkUrl` with `queryIndex`

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 8 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Updated

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 10 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Updated the Requirement

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 11 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: I have added `continue` later and forgot to remove `else` part
Now, Updated the Requirement

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 9 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Updated.
Thank you for the input, I shall use this approach for later projects as well.

**Comment by RamMohanMadhamanchi** [comment]
All the requested changes were updated, please review them

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 12]
```diff
@@ -5,37 +5,44 @@ const LAST_PUBLISHED_VERSION = 6;
...
+const CHAT_LINK_INDEX = 14;
 /**
  * If you add/remove/modify columns, you must also change
  * 1. DATA_END_COLUMN
  * 2. MSG_HELP
  */
 
 const DATA_BEGIN_ROW = 2;
-const DATA_END_COLUMN = "L";
+const DATA_END_COLUMN = "N";
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: Shouldn't the `DATA_END_COLUMN` be `O`?

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 13]
```diff
@@ -54,23 +61,26 @@ Version: ${LAST_PUBLISHED_VERSION + 1}
...
+12. Mentor email - Email id of the mentor. To be filled by the team.
+13. Mentor response - Auto populated by the script.
+14. Recording Link - Auto populated by the script.
+15. Chat Link - Auto populated by the script.
 `;
 
 const DATE_DELIMITER = "-";
 
-const MEETING_TITLE_PREFIX = " - 1:1 (";
+const DEFAULT_SESSION_TYPE = "1:1";
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: There is a naming convention here. The prefix must be `MEETING_`. You can name it as `MEETING_DEFAULT_TYPE`.

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 14]
```diff
@@ -626,24 +635,141 @@ function sendInvitesToMentors() {
...
+      noRecordingLinks.push(rowNumber);
+      noChatLinks.push(rowNumber);
+      continue;
+    }
+    let videoFound = false;
+    let chatFound = false;
+    for (let i = 0; i < eventAttachments.length; ++i) {
+      let currentAttachment = eventAttachments[i];
+      let linkUrl = currentAttachment.fileUrl;
+      if (!currentAttachment || !linkUrl) {
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: 1. Null check should happen before the usage. If `currentAttachment` were null, it would fail in the previous line itself.
2. There is no need of `linkUrl` variable. We essentially use it only twice - once for null check and then for `setValue`. So use it directly.

File: peer-to-peer-scheduler/peer-to-peer-scheduler.gs [hunk 15]
```diff
@@ -0,0 +1,721 @@
+// Whenever you publish a new version of the add-on, you must increment the below version number. This helps for book-keeping and comparing with previous versions.
+const LAST_PUBLISHED_VERSION = 0;
```
Comment (sailesh-ramayanam) [inline peer-to-peer-scheduler/peer-to-peer-scheduler.gs]: Please move P2P scheduler to a separate PR.

**Review by sailesh-ramayanam — CHANGES_REQUESTED** [review]

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 12 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Corrected to `DATA_END_COLUMN` to `O`

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 13 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Changed `DEFAULT_SESSION_TYPE` to `MEETING_DEFAULT_TYPE`

File: one-to-one-scheduler/one-to-one-scheduler.gs (hunk 14 above)
Comment (RamMohanMadhamanchi) [reply one-to-one-scheduler/one-to-one-scheduler.gs]: Removed `linkUrl` and Used `currentAttachment.fileUrl`

File: peer-to-peer-scheduler/peer-to-peer-scheduler.gs (hunk 15 above)
Comment (RamMohanMadhamanchi) [reply peer-to-peer-scheduler/peer-to-peer-scheduler.gs]: Removed P2P Scheduler in this PR. Will add them after the approval of this PR

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 16]
```diff
@@ -282,13 +321,19 @@ function createMeeting(calendar, details) {
     studentName = studentEmail;
   }
 
+  let sessionType = details[SESSION_TYPE_INDEX].trim();
+  if (sessionType.length === 0) {
+    // If Session type is not given, we will take session type as 1:1
+    sessionType = DEFAULT_SESSION_TYPE;
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: This has to be `MEETING_DEFAULT_TYPE`.

File: one-to-one-scheduler/one-to-one-scheduler.gs [hunk 17]
```diff
@@ -626,24 +635,140 @@ function sendInvitesToMentors() {
...
+      chatLinkCell.setValue(NOT_FOUND);
+      noRecordingLinks.push(rowNumber);
+      noChatLinks.push(rowNumber);
+      continue;
+    }
+    let videoFound = false;
+    let chatFound = false;
+    for (let i = 0; i < eventAttachments.length; ++i) {
+      let currentAttachment = eventAttachments[i];
+      if (!currentAttachment.fileUrl) {
```
Comment (sailesh-ramayanam) [inline one-to-one-scheduler/one-to-one-scheduler.gs]: This must be `!currentAttachment || !currentAttachment.fileUrl`.

**Review by sailesh-ramayanam — CHANGES_REQUESTED** [review]

**Review by sailesh-ramayanam — APPROVED** [review]

**Comment by RamMohanMadhamanchi** [comment]
[add-on-1-to-1-test-matrix](https://docs.google.com/spreadsheets/d/1yQMdosFKNQ8_p-DD4Krac_9CEpbfPCUUkziR3Oj2eH0/edit?usp=sharing)
